virtcontainers/cgroups.go

```diff
@@ -0,0 +1,192 @@
+		return nil
+	}
+
+	if err := s.cgroup.sandboxSub.MoveTo(s.cgroup.commonParent); err != nil {
+		return fmt.Errorf("failed to clear cgroup processes")
+	}
+
+	return s.cgroup.sandboxSub.Delete()
+}
+
+func (s *Sandbox) setupCgroups() error {
+	if s.cgroup == nil {
+		return fmt.Errorf("failed to setup uninitialized cgroup for sandbox")
+	}
+
+	resource, err := s.mergeSpecResource()
+	if err != nil {
+		return err
+	}
+
+	if err := s.applyCPUCgroup(resource); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Sandbox) applyCPUCgroup(rc *specs.LinuxResources) error {
+	if s.cgroup == nil {
+		return fmt.Errorf("failed to setup uninitialized cgroup for sandbox")
+	}
+
+	// apply cpu constraint to vcpu cgroup
+	if err := s.cgroup.vcpuSub.Update(rc); err != nil {
+		return err
+	}
+
+	// when new container joins, new CPU could be hotplugged, so we
+	// have to query fresh vcpu info from hypervisor for every time.
+	tids, err := s.hypervisor.getThreadIDs()
+	if err != nil {
+		return err
+	}
+
+	if tids == nil {
+		// TODO: we only support qemu in this scenario, for unsupported hypervisor,
+		// just ignore and print a warning
+		s.Logger().Warning("failed to get thread ids from hypervisor")
+		return nil
+	}
+
+	// use Add() to add vcpu thread to s.cgroup, it will write thread id to
+	// `cgroup.procs` which will move all threads in qemu process to this cgroup
+	// immediately as default behaviour.
+	if len(tids.vcpus) > 0 {
+		if err := s.cgroup.sandboxSub.Add(cgroups.Process{
+			Pid: tids.vcpus[0],
+		}); err != nil {
+			return err
+		}
+	}
+
+	for _, i := range tids.vcpus {
+		if i <= 0 {
+			continue
+		}
+
+		// In contrast, AddTask will write thread id to `tasks`
+		// After this, vcpu threads are in "vcpu" sub-cgroup, other threads in
+		// qemu will be left in parent cgroup untouched.
+		if err := s.cgroup.vcpuSub.AddTask(cgroups.Process{
+			Pid: i,
+		}); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *Sandbox) mergeSpecResource() (*specs.LinuxResources, error) {
+	if s.config == nil {
+		return nil, fmt.Errorf("sandbox config is nil")
+	}
+
+	resource := &specs.LinuxResources{
+		CPU: &specs.LinuxCPU{},
+	}
+
+	for _, c := range s.config.Containers {
+		config, ok := c.Annotations[annotations.ConfigJSONKey]
+		if !ok {
+			s.Logger().WithField("container", c.ID).Warningf("failed to find config from container annotations")
+			continue
+		}
+
+		var spec specs.Spec
+		if err := json.Unmarshal([]byte(config), &spec); err != nil {
+			return nil, err
+		}
+
+		// TODO: how to handle empty/unlimited resource?
+		// maybe we should add a default CPU/Memory delta when no
+		// resource limit is given. -- @WeiZhang555
+		if spec.Linux == nil || spec.Linux.Resources == nil {
+			continue
+		}
+		// calculate cpu quota and period
+		rc := spec.Linux.Resources
+		if rc.CPU != nil && rc.CPU.Quota != nil && rc.CPU.Period != nil &&
+			*rc.CPU.Quota > 0 && *rc.CPU.Period > 0 {
+			if resource.CPU.Period == nil {
```

Comment:
I'd consider putting this logic into a function. That would:

- encapsulate the calculation more clearly.
- allow you to write a unit test specifically for the calculation.
- reduce code complexity.

Reply:
@jodh-intel The function is created, since `TestMergeSpecResource` covers all the cases, I didn't add new unit tests.